File: hotel-reservation-application/src/backend/services/search_service/instrumentation.py
"""
instrumentation.py
OpenTelemetry instrumentation setup for Python Search Service

This module sets up distributed tracing for the FastAPI application.
It can be enabled/disabled via the TRACING_ENABLED environment variable.
"""

import os
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_NAMESPACE, SERVICE_VERSION
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.elasticsearch import ElasticsearchInstrumentor

# Check if tracing is enabled via environment variable
TRACING_ENABLED = os.getenv("TRACING_ENABLED", "false").lower() == "true"


def setup_tracing(app):
    """
    Setup OpenTelemetry tracing for FastAPI application
    
    This function:
    1. Creates a TracerProvider with service metadata
    2. Configures OTLP exporter to send traces to OpenTelemetry Collector
    3. Instruments FastAPI to automatically create spans for HTTP requests
    4. Instruments Elasticsearch client to trace database queries
    
    Args:
        app: FastAPI application instance
    
    Returns:
        None
    """
    if not TRACING_ENABLED:
        print("⚠️  [TRACING] Tracing is DISABLED (TRACING_ENABLED=false)")
        print("⚠️  [TRACING] Set TRACING_ENABLED=true to enable distributed tracing")
        return
    
    print("🔍 [TRACING] Initializing OpenTelemetry for search-service...")
    
    # Create resource with service information
    # This metadata will be attached to all spans
    resource = Resource(attributes={
        SERVICE_NAME: os.getenv("OTEL_SERVICE_NAME", "search-service"),
        SERVICE_NAMESPACE: "hotel-reservation",
        SERVICE_VERSION: "1.0.0",
    })
    
    # Create TracerProvider with the resource
    provider = TracerProvider(resource=resource)
    
    # Create OTLP exporter to send traces to OpenTelemetry Collector
    otlp_endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4318")
    exporter = OTLPSpanExporter(
        endpoint=f"{otlp_endpoint}/v1/traces",
        headers={}
    )
    
    # Add BatchSpanProcessor to batch spans before sending
    # This improves performance by reducing network calls
    processor = BatchSpanProcessor(exporter)
    provider.add_span_processor(processor)
    
    # Set the global TracerProvider
    trace.set_tracer_provider(provider)
    
    # Instrument FastAPI application
    # This automatically creates spans for all HTTP requests
    FastAPIInstrumentor.instrument_app(app)
    
    # Instrument Elasticsearch client
    # This automatically creates spans for all Elasticsearch queries
    ElasticsearchInstrumentor().instrument()
    
    print("✅ [TRACING] OpenTelemetry initialized successfully")
    print(f"📡 [TRACING] Exporting traces to: {otlp_endpoint}")
    print("✅ [TRACING] Auto-instrumentation enabled for: FastAPI, Elasticsearch")

File: hotel-reservation-application/src/backend/services/nodejs-bff/instrumentation.js
// instrumentation.js
// OpenTelemetry instrumentation for Node.js BFF service
// 
// IMPORTANT: This file MUST be imported BEFORE any other imports in server.js
// This ensures that all HTTP/Express/Axios libraries are instrumented before they are used

const { NodeSDK } = require('@opentelemetry/sdk-node');
const { getNodeAutoInstrumentations } = require('@opentelemetry/auto-instrumentations-node');
const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-http');
const { Resource } = require('@opentelemetry/resources');
const { SemanticResourceAttributes } = require('@opentelemetry/semantic-conventions');

// Check if tracing is enabled via environment variable
const TRACING_ENABLED = process.env.TRACING_ENABLED === 'true';

if (TRACING_ENABLED) {
  console.log('🔍 [TRACING] Initializing OpenTelemetry for nodejs-bff...');
  console.log(`📡 [TRACING] Exporting to: ${process.env.OTEL_EXPORTER_OTLP_ENDPOINT}`);

  // Create OpenTelemetry SDK
  const sdk = new NodeSDK({
    // Resource identifies this service in traces
    resource: new Resource({
      [SemanticResourceAttributes.SERVICE_NAME]: process.env.OTEL_SERVICE_NAME || 'nodejs-bff',
      [SemanticResourceAttributes.SERVICE_NAMESPACE]: 'hotel-reservation',
      [SemanticResourceAttributes.SERVICE_VERSION]: '1.0.0',
    }),
    
    // OTLP exporter sends traces to OpenTelemetry Collector
    traceExporter: new OTLPTraceExporter({
      url: process.env.OTEL_EXPORTER_OTLP_ENDPOINT + '/v1/traces',
      headers: {},
    }),
    
    // Auto-instrumentations for common libraries
    instrumentations: [
      getNodeAutoInstrumentations({
        // HTTP/HTTPS instrumentation (incoming and outgoing requests)
        '@opentelemetry/instrumentation-http': {
          enabled: true,
          // Ignore health check endpoints
          ignoreIncomingRequestHook: (req) => {
            return req.url === '/health';
          },
        },
        // Express.js instrumentation
        '@opentelemetry/instrumentation-express': {
          enabled: true,
        },
        // Axios instrumentation (HTTP client for calling search-service)
        '@opentelemetry/instrumentation-axios': {
          enabled: true,
        },
      }),
    ],
  });

  // Start the SDK
  sdk.start();
  console.log('✅ [TRACING] OpenTelemetry initialized successfully');
  console.log('✅ [TRACING] Auto-instrumentation enabled for: HTTP, Express, Axios');

  // Graceful shutdown on process termination
  process.on('SIGTERM', () => {
    sdk.shutdown()
      .then(() => console.log('🛑 [TRACING] OpenTelemetry shut down successfully'))
      .catch((error) => console.error('❌ [TRACING] Error shutting down OpenTelemetry', error))
      .finally(() => process.exit(0));
  });
} else {
  console.log('⚠️  [TRACING] Tracing is DISABLED (TRACING_ENABLED=false)');
  console.log('⚠️  [TRACING] Set TRACING_ENABLED=true to enable distributed tracing');
}

module.exports = { TRACING_ENABLED };

